### PR TITLE
nixos/matrix-synapse: update docs

### PR DESCRIPTION
**Pull request by Ma27**
###### Description of changes

* Update attribute names in code examples (* -> settings.*).
* Use `nix-shell -p` rather than `nix run` because the example won't
  work with the current default Nix.
* Update config values for `element-web`.
* Fix link to `element-web` security considerations.
* Make the synapse expression even smaller and use callout-lists to
  explain the code.
* Document how to correctly deploy the shared registration secret.

[1] https://spec.matrix.org/latest/client-server-api/#getwell-knownmatrixclient

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by mweinelt** [comment]
Synapse has a new config documentation that we should reference in the settings option. 
https://matrix-org.github.io/synapse/latest/usage/configuration/config_documentation.html

**Comment by dali99** [comment]
Do note that upstream recommends running everything through a reverse proxy https://matrix-org.github.io/synapse/latest/reverse_proxy.html with something like nginx doing the SSL-termination for both federation and client

File: nixos/modules/services/matrix/synapse.xml [hunk 1]
```diff
@@ -33,18 +33,28 @@
...
+    add_header Access-Control-Allow-Origin *;
+    return 200 '${builtins.toJSON data}';
+  '';
 in {
   networking = {
     <link linkend="opt-networking.hostName">hostName</link> = "myhostname";
     <link linkend="opt-networking.domain">domain</link> = "example.org";
   };
-  <link linkend="opt-networking.firewall.allowedTCPPorts">networking.firewall.allowedTCPPorts</link> = [ 80 443 ];
+  <link linkend="opt-networking.firewall.allowedTCPPorts">networking.firewall.allowedTCPPorts</link> = [ 80 443 8448 ];
```
Comment (SuperSandro2000) [inline nixos/modules/services/matrix/synapse.xml:57]: Isn't it pretty common to use 443 and a reverse proxy? I think we should propose that in the docs.

**Review by SuperSandro2000 — COMMENTED** [review]
Can we convert the doc to markdown?

**Comment by Ma27** [comment]
Updated accordingly.

Also demonstrated how to not store the registration secret in the store and decided to use callout lists to explain parts of the synapse expression.

File: nixos/modules/services/matrix/synapse.xml [hunk 2]
```diff
@@ -57,99 +62,95 @@ in {
...
+     of <package>matrix-synapse</package> is equal to the domain (i.e.
+     <literal>example.org</literal> from <literal>@foo:example.org</literal>)
+     and the federation port is 8448.
+     Further reference can be found in the <link xlink:href="https://matrix-org.github.io/synapse/latest/delegate.html">docs
+     about delegation</link>.
+    </para>
+   </callout>
+   <callout arearefs='ex-matrix-synapse-well-known-client'>
+    <para>
+     This is usually needed for homeserver discovery (from e.g. another server).
```
Comment (dali99) [inline nixos/modules/services/matrix/synapse.xml:133]: The biggest deal here is to enable logging in with matrix ids. it's used by clients, not by other servers

File: nixos/modules/services/matrix/synapse.xml [hunk 3]
```diff
@@ -57,99 +62,95 @@ in {
...
+     Further reference can be found in the <link xlink:href="https://spec.matrix.org/latest/client-server-api/#getwell-knownmatrixclient">upstream docs</link>
+    </para>
+   </callout>
+   <callout arearefs='ex-matrix-synapse-rev-default'>
+    <para>
+     It's also possible to do a redirect here or something else, this vhost is not
+     needed for Matrix. It's recommended though to <emphasis>not put</emphasis> element
+     here, see also the <link linkend='ex-matrix-synapse-rev-default'>section about Element</link>.
```
Comment (dali99) [inline nixos/modules/services/matrix/synapse.xml:141]: Maybe add that this is for security reasons, CORS headers require different domains for proper protection

Comment (dali99) [reply nixos/modules/services/matrix/synapse.xml:141]: Oh you did never mind

**Comment by Ma27** [comment]
ping @dali99

**Review by D4ndellion — CHANGES_REQUESTED** [review]
Looks good for a base configuration!

Only thing is that you should also proxy `/_synapse/client` since it is used for things like password resets and SSO

**Review by D4ndellion — APPROVED** [review]

**Pull request merged**